valet/valet_suite_test.go

```diff
@@ -37,252 +38,246 @@ import (
-			foundDirs = append(foundDirs, p)
+		for path := range paths {
+			foundDirs = append(foundDirs, path)
 		}
-
-		select {
-		case err := <-errs:
+		for err := range errs {
 			Expect(err).NotTo(HaveOccurred())
-		default:
 		}
 
 		cancel()
 	})
 
-	Context("Using a prune function", func() {
+	When("using a prune function", func() {
 		It("should find paths, except those pruned", func() {
-			Expect(len(foundDirs)).To(Equal(len(expectedPaths)))
-
-			for i, ep := range expectedPaths {
-				a, err := filepath.Abs(ep)
-				Expect(err).NotTo(HaveOccurred())
-
-				fp, _ := valet.NewFilePath(a)
-
-				Expect(foundDirs[i].Location).To(Equal(fp.Location))
-				Expect(foundDirs[i].Info).ToNot(BeNil())
+			expectedPaths := []string{
+				".",
+				"1",
+				"testdir",
 			}
+
+			Expect(foundDirs).To(WithTransform(pathTransform,
+				ConsistOf(expectedPaths)))
 		})
 	})
 })
 
 var _ = Describe("Find files at intervals", func() {
 	var (
+		foundFiles    []valet.FilePath
+		pathTransform localPathTransform
+
+		dataDir = "testdata/valet"
+
 		expectedPaths = []string{
-			"testdata/valet/1/reads/fast5/reads1.fast5",
-			"testdata/valet/1/reads/fast5/reads1.fast5.md5",
-			"testdata/valet/1/reads/fast5/reads2.fast5",
-			"testdata/valet/1/reads/fast5/reads3.fast5",
-			"testdata/valet/1/reads/fastq/reads1.fastq",
-			"testdata/valet/1/reads/fastq/reads1.fastq.md5",
-			"testdata/valet/1/reads/fastq/reads2.fastq",
-			"testdata/valet/1/reads/fastq/reads2.fastq.gz",
-			"testdata/valet/1/reads/fastq/reads3.fastq",
+			"1/reads/fast5/reads1.fast5",
+			"1/reads/fast5/reads1.fast5.md5",
+			"1/reads/fast5/reads2.fast5",
+			"1/reads/fast5/reads3.fast5",
+			"1/reads/fastq/reads1.fastq",
+			"1/reads/fastq/reads1.fastq.md5",
+			"1/reads/fastq/reads2.fastq",
+			"1/reads/fastq/reads2.fastq.gz",
+			"1/reads/fastq/reads3.fastq",
 		}
-
-		foundFiles = map[string]bool{}
 	)
 
 	BeforeEach(func() {
+		absDataDir, err := filepath.Abs(dataDir)
+		Expect(err).NotTo(HaveOccurred())
+		pathTransform = makeLocalPathTransform(absDataDir)
+
 		cancelCtx, cancel := context.WithCancel(context.Background())
 		interval := 500 * time.Millisecond
 
-		paths, errs := valet.FindFilesInterval(cancelCtx, "testdata/valet",
+		paths, errs := valet.FindFilesInterval(cancelCtx, dataDir,
 			valet.IsRegular, valet.IsFalse, interval)
 
 		// Find files or timeout and cancel
-		done := make(chan bool, 2)
+		found := make(map[string]valet.FilePath) // FilePaths are not comparable
+
+		var wg sync.WaitGroup
+		wg.Add(1)
 
 		go func() {
+			defer wg.Done()
 			defer cancel()
 
 			timer := time.NewTimer(5 * interval)
-			<-timer.C
-			done <- true // Timeout
-		}()
 
-		go func() {
-			defer cancel()
-
-			foundFiles = make(map[string]bool) // FilePaths are not comparable
-			for p := range paths {
-				foundFiles[p.Location] = true
-				if len(foundFiles) >= len(expectedPaths) {
-					done <- true // Find files
+			for {
+				select {
+				case <-timer.C:
 					return
```

Comment:
Can this ever happen? Doesn't it remain in the paths loop?

Reply:
Yes, it returns, either when the timer pings, or when sufficient file have been found that the test might pass before then. I am leaking timers here, though - I think I'm looking for `time.After`

Reply:
Yes, you can use a time.After but will still need to timer.Stop() before returning.
What I didn't  and still don't get is the behaviour of your loop. https://play.golang.org/p/rWCVkwqxChb shows that once you start ranging over your paths channel, it will stay in that range loop until the channel is closed, and the timer case won't trigger. Yet somehow your loops do work. Can you explain what's going on?

Reply:
I think this has the exact same behaviour as what you have:
```
                 go func() {
			defer wg.Done()
			defer cancel()

			for path := range paths {
				found[path.Location] = path
				if len(found) >= len(expectedPaths) {
					// Find files
					return
				}
			}
		}()
```

Reply:
Yes, I see now. Thanks - I need to fix that.